### PR TITLE
I1319 middleware check percentage field

### PR DESCRIPTION
**Pull request by Serhiy91**

### CODE REVIEW

File: share/middleware/document/checkPercentField.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { compose, sum, map, prop } from 'ramda';
+import invariant from 'invariant';
+
+import { MAX_TOTAL_PERCENT } from '../../constants';
+import Errors from '../../errors';
+
+export default ({ collection, key }) => async (next, root, args, context) => {
```
Comment (Dartv) [inline share/middleware/document/checkPercentField.js]: use function instead and don't hardcode collection as string, use context instead

File: share/middleware/document/checkPercentField.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  const { [key]: percent = 0, organizationId } = root || {};
+  const { [key]: newPercent } = args;
+  const { collections: { [collection]: col } } = context;
+
+  const percentDiff = newPercent - percent;
```
Comment (Dartv) [inline share/middleware/document/checkPercentField.js]: newPercent can be undefined, then percentDiff is `NaN`, then `percentDiff <= 0` is `false`

File: share/middleware/document/checkPercentField.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  if (percentDiff <= 0) return next(root, args, context);
+
+  const items = col.find({ organizationId }).fetch();
```
Comment (Dartv) [inline share/middleware/document/checkPercentField.js]: better to use `query` returned from config function like all other middleware, but default to `organizationId`

File: share/middleware/document/checkPercentField.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  const totalPercent = compose(sum, map(prop(key)))(items);
```
Comment (Dartv) [inline share/middleware/document/checkPercentField.js]: make a function at the top of the file. Something like `calculateTotalPercent`

File: share/middleware/document/checkPercentField.js (hunk 3 above)
Comment (Dartv) [reply share/middleware/document/checkPercentField.js]: Also, use `fields` here for efficiency.

File: share/middleware/document/checkPercentField.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { compose, sum, map, prop } from 'ramda';
```
Comment (Dartv) [inline share/middleware/document/checkPercentField.js]: Make tests for this middleware

File: share/errors.js [hunk 6]
```diff
@@ -30,4 +30,5 @@ export default {
   ACT_CANNOT_SET_COMPLETION_DATE_FOR_COMPLETED: 'Cannot set completion date for completed action',
   ACT_CANNOT_SET_EXECUTOR_FOR_COMPLETED: 'Cannot set executor for completed action',
   ST_CANNOT_CHANGE_DEFAULT: 'Cannot change or delete the default standard type',
+  EXCEEDED_TOTAL_AMOUNT_OF_PERCENT: 'The total amount of percent is exceeded 100%',
```
Comment (Dartv) [inline share/errors.js]: Steve's proposed error message: "Cost lines cannot add up to more than 100% Please reduce one or more values".
I propose that we allow customizable error in middleware, we can specify collection name in config.

**Comment by Serhiy91** [comment]
@Dartv fixed

**Pull request merged**